plugins/hls-explicit-record-fields-plugin/src/Ide/Plugin/ExplicitFields.hs

```diff
@@ -0,0 +1,266 @@
+  nfp <- getNormalizedFilePath (docId ^. L.uri)
+  pragma <- getFirstPragma pId ideState nfp
+  CRR renderedRecs (map unExt -> exts) <- collectRecordsInRange range ideState nfp
+  let actions = map (mkCodeAction nfp exts pragma) renderedRecs
+  pure $ List actions
+
+  where
+    mkCodeAction :: NormalizedFilePath -> [Extension] -> NextPragmaInfo -> RenderedRecordInfo -> Command |? CodeAction
+    mkCodeAction nfp exts pragma rec = InR CodeAction
+      { _title = mkCodeActionTitle exts
+      , _kind = Just CodeActionRefactorRewrite
+      , _diagnostics = Nothing
+      , _isPreferred = Nothing
+      , _disabled = Nothing
+      , _edit = Just $ mkWorkspaceEdit nfp edits
+      , _command = Nothing
+      , _xdata = Nothing
+      }
+      where
+        edits = catMaybes [ mkTextEdit rec , pragmaEdit ]
+
+        mkTextEdit :: RenderedRecordInfo -> Maybe TextEdit
+        mkTextEdit (RenderedRecordInfo ss r) = TextEdit <$> srcSpanToRange ss <*> pure r
+
+        -- NOTE(ozkutuk): `RecordPuns` extension is renamed to `NamedFieldPuns`
+        -- in GHC 9.4, but we still want to insert `NamedFieldPuns` in pre-9.4
+        -- GHC as well, hence the replacement.
+        -- https://gitlab.haskell.org/ghc/ghc/-/merge_requests/6156
+        pragmaEdit :: Maybe TextEdit
+        pragmaEdit = if NamedFieldPuns `elem` exts
+                       then Nothing
+                       else Just $ patchExtName $ insertNewPragma pragma NamedFieldPuns
+          where
+            patchExtName = L.newText %~ T.replace "Record" "NamedField"
+
+    mkWorkspaceEdit :: NormalizedFilePath -> [TextEdit] -> WorkspaceEdit
+    mkWorkspaceEdit nfp edits = WorkspaceEdit changes Nothing Nothing
+      where
+        changes = Just $ HashMap.singleton (fromNormalizedUri (normalizedFilePathToUri nfp)) (List edits)
+
+    mkCodeActionTitle :: [Extension] -> Text
+    mkCodeActionTitle exts =
+      if NamedFieldPuns `elem` exts
+        then title
+        else title <> " (needs extension: NamedFieldPuns)"
+        where
+          title = "Expand record wildcard"
+
+collectRecordsRule :: Recorder (WithPriority Log) -> Rules ()
+collectRecordsRule recorder = define (cmapWithPrio LogShake recorder) $ \CollectRecords nfp -> do
+  tmr <- use TypeCheck nfp
+  let exts = getEnabledExtensions <$> tmr
+      recs = getRecords <$> tmr
+      renderedRecs = mapMaybe renderRecordInfo <$> recs
+  pure ([], CRR <$> renderedRecs <*> exts)
+
+getEnabledExtensions :: TcModuleResult -> [GhcExtension]
+getEnabledExtensions = map GhcExtension . toList . extensionFlags . ms_hspp_opts . pm_mod_summary . tmrParsed
+
+getRecords :: TcModuleResult -> [RecordInfo]
+getRecords (tmrRenamed -> (hs_valds -> valBinds,_,_,_)) =
+  collectRecords valBinds
+
+data CollectRecords = CollectRecords
+                    deriving (Eq, Show, Generic)
+
+instance Hashable CollectRecords
+instance NFData CollectRecords
+
+data CollectRecordsResult = CRR
+  { recordInfos       :: ![RenderedRecordInfo]
```

Comment:
Making a field of type `[a]` strict typically does very little: all it does is force the outermost cons constructor, while the rest of the spine of the list will not be forced. If you want the container to be strict, do this and use a `Vector` or similar.

Reply:
I made it strict out of habit, I don't think we necessarily need to make this container strict. I will remove the strictness annotation to avoid future confusion.

Reply:
I usually just reach for `StrictData` for that use case. Here it stands out, which made me think it might have a specific intention behind it!